File: packages/vue-im-ui/src/shared/contracts/form-behavior.ts
export type FormControlState = "default" | "hover" | "focus" | "pressed" | "filled" | "invalid" | "disabled" | "readOnly" | "loading";
export type FormKeyboardIntent = "none" | "tabForward" | "tabBackward" | "submit" | "closePopup" | "nextOption" | "previousOption" | "firstOption" | "lastOption" | "pageForward" | "pageBackward";

export interface FormKeyboardInput {
  key: string;
  shift?: boolean;
  composing?: boolean;
  popupOpen?: boolean;
  submitOnEnter?: boolean;
}

/**
 * One rule for every desktop key a form control sees.
 *
 * While an IME composition is open the IME owns the keyboard — Enter commits a
 * candidate, Escape cancels the composition, Tab and the arrows walk the
 * candidate window. So composing short-circuits everything: the control must not
 * submit, close its popup, or move focus out from under the user mid-word.
 */
export function resolveFormKeyboardIntent(input: FormKeyboardInput): FormKeyboardIntent {
  if (input.composing) return "none";
  const key = input.key.toLowerCase();
  if (key === "tab") return input.shift ? "tabBackward" : "tabForward";
  if (key === "escape" && input.popupOpen) return "closePopup";
  if (key === "enter" && input.submitOnEnter) return "submit";
  if (!input.popupOpen) return "none";
  if (key === "arrowdown") return "nextOption";
  if (key === "arrowup") return "previousOption";
  if (key === "home") return "firstOption";
  if (key === "end") return "lastOption";
  if (key === "pagedown") return "pageForward";
  if (key === "pageup") return "pageBackward";
  return "none";
}

File: packages/ios-im-ui/Sources/FlareIMUI/Models/FormBehavior.swift
public enum FlareFormControlState: String, Sendable { case idle, hover, focus, pressed, filled, invalid, disabled, readOnly, loading }
public enum FlareFormKeyboardIntent: String, Sendable { case none, tabForward, tabBackward, submit, closePopup, nextOption, previousOption, firstOption, lastOption, pageForward, pageBackward }

public func resolveFormKeyboardIntent(
    _ key: String,
    shift: Bool = false,
    composing: Bool = false,
    popupOpen: Bool = false,
    submitOnEnter: Bool = false
) -> FlareFormKeyboardIntent {
    // While an IME composition is open the IME owns the keyboard — Enter commits a
    // candidate, Escape cancels the composition, Tab and the arrows walk the
    // candidate window. Composing short-circuits everything.
    if composing { return .none }
    let key = key.lowercased()
    if key == "tab" { return shift ? .tabBackward : .tabForward }
    if key == "escape" && popupOpen { return .closePopup }
    if key == "enter" && submitOnEnter { return .submit }
    if !popupOpen { return .none }
    switch key {
    case "arrowdown": return .nextOption
    case "arrowup": return .previousOption
    case "home": return .firstOption
    case "end": return .lastOption
    case "pagedown": return .pageForward
    case "pageup": return .pageBackward
    default: return .none
    }
}
